lib/streamlit/uploaded_file_manager.py

```diff
@@ -14,20 +14,32 @@
 
 import io
 import threading
-from typing import Dict
-from typing import List
-from typing import NamedTuple
-from typing import Tuple
+from typing import Dict, NamedTuple, Optional, List, Tuple
 from blinker import Signal
 
 
+class UploadedFileRec(NamedTuple):
+    """Metadata and raw bytes for an uploaded file. Immutable."""
+
+    id: str
+    name: str
+    type: str
+    data: bytes
+
+
 class UploadedFile(io.BytesIO):
-    def __init__(self, id, name, type, data, **kwargs):
-        super(UploadedFile, self).__init__(data)
-        self.id = id
-        self.name = name
-        self.type = type
-        self.size = self.getbuffer().nbytes
```

Comment:
This is an important but subtle change: `self.getbuffer()` returns a _copy_ of the underlying buffer. Instead, we just do `len(data)`, which doesn't allocate.

Reply:
yea I did not like that. I tried out a bunch of things that looked like it was giving me the wrong size but looking at it now it's still off. Looks like it's the frontend that was off 🤦‍♀️   Issue #2431 created for this